### PR TITLE
Do not ignore error in os_free() when size == 0

### PR DESCRIPTION
**Pull request by ldorau**
### Description

Do not ignore error in `os_free()` when size == 0.

Ref: #481

### Checklist
- [x] Code compiles without errors locally
- [x] All tests pass locally
- [x] CI workflows execute properly

### CODE REVIEW

File: benchmark/ubench.c [hunk 1]
```diff
@@ -196,6 +196,7 @@ UBENCH_EX(simple, proxy_pool_with_os_memory_provider) {
     }
 
     umf_memory_pool_handle_t proxy_pool;
+    // the proxy pool requires the tracking provider to work correctly
```
Comment (vinser52) [inline benchmark/ubench.c:199]: The tracking provider is always required to work correctly :)
Would you like to say that tracking capabilities should be ON?

My general thoughts: the amount of features that depend on tracking capabilities is increasing. We should consider making the tracking provider always ON (remove the option that allows to disable it) or we need to check it at build/run time and return appropriate error. I vote for the first option (always unconditional ON).

File: benchmark/ubench.c [hunk 2]
```diff
@@ -238,8 +239,10 @@ UBENCH_EX(simple, disjoint_pool_with_os_memory_provider) {
     disjoint_memory_pool_params.MinBucketSize = DISJOINT_POOL_MIN_BUCKET_SIZE;
 
     umf_memory_pool_handle_t disjoint_pool;
-    umf_result = umfPoolCreate(umfDisjointPoolOps(), os_memory_provider,
-                               &disjoint_memory_pool_params, 0, &disjoint_pool);
+    // disable the tracking provider to make it work faster (UMF_POOL_CREATE_FLAG_DISABLE_TRACKING)
```
Comment (vinser52) [inline benchmark/ubench.c:242]: Are we interested in performance without a tracking provider? But in real workloads tracking capabilities will be ON?

File: src/provider/provider_tracking.c [hunk 3]
```diff
@@ -343,6 +348,15 @@ static umf_result_t trackingFree(void *hProvider, void *ptr, size_t size) {
         umf_ba_global_free(value);
     }
 
+    // umfMemoryProviderFree() should not be called with size == 0,
```
Comment (vinser52) [inline src/provider/provider_tracking.c:351]: But according to this PR  `umfMemoryProviderFree()` for tracking provider can be called with `size == 0`.

Maybe a particular pool implementation should explicitly get the size from the tracker and call `umfMemoryProviderFree()` with the right size?

**Review by vinser52 — CHANGES_REQUESTED** [review]

File: benchmark/ubench.c (hunk 1 above)
Comment (ldorau) [reply benchmark/ubench.c:199]: I agree. Done: https://github.com/oneapi-src/unified-memory-framework/pull/485

File: benchmark/ubench.c (hunk 2 above)
Comment (ldorau) [reply benchmark/ubench.c]: Right. Removed.
Done.

File: src/provider/provider_tracking.c (hunk 3 above)
Comment (ldorau) [reply src/provider/provider_tracking.c:351]: Right. New fix: https://github.com/oneapi-src/unified-memory-framework/pull/487

**Comment by ldorau** [comment]
Replaced with #487

**Pull request closed without merging**